### ISSUE TITLE
Apple's debug symbol extraction tool dsymutil shows warnings for users of libjpeg-turbo static libraries

### ISSUE THREAD

**Issue by alcroito** [libjpeg-turbo/libjpeg-turbo#754]
**Have you searched the existing issues (both open and closed) in the libjpeg-turbo issue tracker to ensure that this bug report is not a duplicate?**
yes

**Does this bug report describe one of the [two known and unsolvable issues with the JPEG format](https://libjpeg-turbo.org/pmwiki/uploads/About/TwoIssueswiththeJPEGStandard.pdf)?**
no

**Clear and concise description of the bug:**
The turbojpeg-static static archive includes object files that have duplicate names, due to including both `$<TARGET_OBJECTS:jpeg16-static>` and `$<TARGET_OBJECTS:jpeg12-static>`, where each contain a subset of common files.

This causes various warnings when running Apple's debug extraction tool dsymutil on shared libraries that use the turbojpeg-static archive:
```
$ dsymutil libmyturbojpeg.dylib
warning: (arm64)  skipping debug map object with duplicate name and timestamp: 2024-03-15 13:54:08.000000000 /Users/alex/Dev/unsorted/libjpeg-turbo/build/libturbojpeg.a(jcapistd.c.o)
warning: (arm64)  skipping debug map object with duplicate name and timestamp: 2024-03-15 13:54:08.000000000 /Users/alex/Dev/unsorted/libjpeg-turbo/build/libturbojpeg.a(jccolor.c.o)
warning: (arm64)  skipping debug map object with duplicate name and timestamp: 2024-03-15 13:54:08.000000000 /Users/alex/Dev/unsorted/libjpeg-turbo/build/libturbojpeg.a(jcdiffct.c.o)
warning: (arm64)  skipping debug map object with duplicate name and timestamp: 2024-03-15 13:54:08.000000000 /Users/alex/Dev/unsorted/libjpeg-turbo/build/libturbojpeg.a(jclossls.c.o)
warning: (arm64)  skipping debug map object with duplicate name and timestamp: 2024-03-15 13:54:08.000000000 /Users/alex/Dev/unsorted/libjpeg-turbo/build/libturbojpeg.a(jcmainct.c.o)
warning: (arm64)  skipping debug map object with duplicate name and timestamp: 2024-03-15 13:54:08.000000000 /Users/alex/Dev/unsorted/libjpeg-turbo/build/libturbojpeg.a(jcprepct.c.o)
warning: (arm64)  skipping debug map object with duplicate name and timestamp: 2024-03-15 13:54:08.000000000 /Users/alex/Dev/unsorted/libjpeg-turbo/build/libturbojpeg.a(jcsample.c.o)
warning: (arm64)  skipping debug map object with duplicate name and timestamp: 2024-03-15 13:54:08.000000000 /Users/alex/Dev/unsorted/libjpeg-turbo/build/libturbojpeg.a(jdapistd.c.o)
warning: (arm64)  skipping debug map object with duplicate name and timestamp: 2024-03-15 13:54:08.000000000 /Users/alex/Dev/unsorted/libjpeg-turbo/build/libturbojpeg.a(jdcolor.c.o)
warning: (arm64)  skipping debug map object with duplicate name and timestamp: 2024-03-15 13:54:08.000000000 /Users/alex/Dev/unsorted/libjpeg-turbo/build/libturbojpeg.a(jddctmgr.c.o)
warning: (arm64)  skipping debug map object with duplicate name and timestamp: 2024-03-15 13:54:08.000000000 /Users/alex/Dev/unsorted
/libjpeg-turbo/build/libturbojpeg.a(jddiffct.c.o)
warning: (arm64)  skipping debug map object with duplicate name and timestamp: 2024-03-15 13:54:08.000000000 /Users/alex/Dev/unsorted/libjpeg-turbo/build/libturbojpeg.a(jdlossls.c.o)
warning: (arm64)  skipping debug map object with duplicate name and timestamp: 2024-03-15 13:54:08.000000000 /Users/alex/Dev/unsorted/libjpeg-turbo/build/libturbojpeg.a(jdmainct.c.o)
warning: (arm64)  skipping debug map object with duplicate name and timestamp: 2024-03-15 13:54:08.000000000 /Users/alex/Dev/unsorted/libjpeg-turbo/build/libturbojpeg.a(jdmerge.c.o)
warning: (arm64)  skipping debug map object with duplicate name and timestamp: 2024-03-15 13:54:08.000000000 /Users/alex/Dev/unsorted/libjpeg-turbo/build/libturbojpeg.a(jdpostct.c.o)
warning: (arm64)  skipping debug map object with duplicate name and timestamp: 2024-03-15 13:54:08.000000000 /Users/alex/Dev/unsorted/libjpeg-turbo/build/libturbojpeg.a(jdsample.c.o)
warning: (arm64)  skipping debug map object with duplicate name and timestamp: 2024-03-15 13:54:08.000000000 /Users/alex/Dev/unsorted/libjpeg-turbo/build/libturbojpeg.a(jidctint.c.o)
warning: (arm64)  skipping debug map object with duplicate name and timestamp: 2024-03-15 13:54:08.000000000 /Users/alex/Dev/unsorted/libjpeg-turbo/build/libturbojpeg.a(jidctred.c.o)
warning: (arm64)  skipping debug map object with duplicate name and timestamp: 2024-03-15 13:54:08.000000000 /Users/alex/Dev/unsorted/libjpeg-turbo/build/libturbojpeg.a(jquant1.c.o)
warning: (arm64)  skipping debug map object with duplicate name and timestamp: 2024-03-15 13:54:08.000000000 /Users/alex/Dev/unsorted/libjpeg-turbo/build/libturbojpeg.a(jquant2.c.o)
warning: (arm64)  skipping debug map object with duplicate name and timestamp: 2024-03-15 13:54:08.000000000 /Users/alex/Dev/unsorted/libjpeg-turbo/build/libturbojpeg.a(jutils.c.o)
warning: (arm64)  skipping debug map object with duplicate name and timestamp: 2024-03-15 12:09:02.000000000 /Users/alex/Dev/unsorted/libjpeg-turbo/build/libturbojpeg.a(rdppm.c.o)
warning: (arm64)  skipping debug map object with duplicate name and timestamp: 2024-03-15 12:09:02.000000000 /Users/alex/Dev/unsorted/libjpeg-turbo/build/libturbojpeg.a(wrppm.c.o)
```

**Steps to reproduce the bug (using *only* libjpeg-turbo):**
```bash
$ cat CMakeLists.txt
$ cmake . -DCMAKE_BUILD_TYPE=Debug && cmake --build .
```
```cmake
cmake_minimum_required(VERSION 3.16)
project(repro LANGUAGES C)

set(lib_source_path "${CMAKE_CURRENT_BINARY_DIR}/lib.c")
file(WRITE "${lib_source_path}" "int f() { return 42; }")
add_library(mylib SHARED "${lib_source_path}")
target_link_libraries(mylib PRIVATE /Users/alex/Dev/unsorted/libjpeg-turbo/build/libturbojpeg.a)
# Just ensures that all symbols are included in the final binary, but this happens with regular
# c code without the flag as well.
target_link_options(mylib PRIVATE -all_load)
add_custom_target(run_dsymutil ALL
  COMMAND dsymutil $<TARGET_FILE:mylib>
  )
add_dependencies(run_dsymutil mylib)
```

**Expected behavior:**
No warnings are shown when running `dsymutil`

**Observed behavior:**
Warnings are shown.

**Platform(s) (compiler version, operating system version, CPU) on which the bug was observed:**
Apple macOS Sonoma 14.3.1 (23D60) (arm64)
Apple clang version 15.0.0 (clang-1500.1.0.2.5) Xcode 15.2
cmake version 3.28.3

**libjpeg-turbo release(s), commit(s), or branch(es) in which the bug was observed (always test the tip of the main branch or the latest [stable pre-release](https://libjpeg-turbo.org/DeveloperInfo/PreReleases) to verify that the bug hasn't already been fixed):**
v3.0.2 tag

**If the bug is a regression, the specific commit that introduced the regression (use `git bisect` to determine this):**
d491094b41220a3eb5694311f11b855ad1300c2e

**Additional information:**
I have an almost complete solution at https://github.com/alcroito/libjpeg-turbo/commit/fcaf30bf243fa58aeddb1480d208b8ebcacf8884

It changes the build system to copy the files at cmake configure time to create unique source file names, and thus object files names, which prevents the warnings. There is a downside described in the commit message. Let me know if this would be an acceptable solution.

Originally encountered while building and using the library downstream for Qt https://bugreports.qt.io/browse/QTBUG-123324

**Comment by dcommander** [comment]
I can't reproduce the issue upstream.  I built libjpeg-turbo with debug symbols and ran `dsymutil` on **tjbench-static**, which links with the static TurboJPEG API library.  I also tried running `dsymutil` on a dynamic library (also built with debug symbols) that links with the static TurboJPEG API library (specifically the TurboVNC Helper library in [TurboVNC](https://github.com/TurboVNC/turbovnc).)  No warnings in either case.  That may be because the issue is only reproducible with a newer Xcode version than I am able to run.  (My test system is a 10-year-old Mac Mini, so it can't be upgraded beyond Monterrey, which means it can't run Xcode 14.3 or later.)

Even if I could reproduce the issue, I'm not sure what to do about it.  We aren't doing anything illegal.  We're just compiling the same source file with different macros defined, which causes different versions of the same functions (with different function names) to be generated for different sample sizes.  Other projects do that as well.  Are these just warnings, or is there some loss of expected functionality?  If the warnings are truly innocuous, then my strong inclination is to do nothing.  IMHO, it isn't really correct for dsymutil to assume that object files with the same name and timestamp are identical.  It should be looking at whether those object files have the same symbols.  (I wonder if there might even be a dsymutil option that tells it to do that.)

Copying/renaming the source files may work for a downstream build system, but it's not an appropriate solution for the upstream build system that is used actively for libjpeg-turbo development.  (It has been my experience that dependencies don't reliably propagate when you copy source files, even if you use CMake's own `copy_if_different` option in a custom command.)  `AVOID_DSYMUTIL_WARNINGS` isn't something that I would enable by default, so it would effectively be cruft (a feature that is never tested upstream and is thus very likely to break as the code evolves.)  In order to avoid cruft, our project does not include build system modifications that are only useful to downstream projects.

I would consider a fix for this issue only if all of the following are true:
1. The issue represents a genuine loss of functionality, i.e. the warnings are not innocuous.
2. There is no way to work around the issue in any other way, e.g. with a dsymutil option.
3. The issue can be reproduced in an unmodified build of libjpeg-turbo.

The way I would fix it would be to introduce wrapper source files that include the actual source files multiple times, e.g.:

**jcapistd.c**
```
#define BITS_IN_JSAMPLE  8
#include "jcapistd-mp.c"
#undef BITS_IN_JSAMPLE
#define BITS_IN_JSAMPLE  12
#include "jcapistd-mp.c"
#undef BITS_IN_JSAMPLE
#define BITS_IN_JSAMPLE  16
#include "jcapistd-mp.c"
#undef BITS_IN_JSAMPLE
```

Note that **turbojpeg.c** already does something like that.  The reason I didn't do the same thing with the libjpeg source files is because there are so many of them, and I didn't want to further clutter the already cluttered source directory.  I still don't, so I would still prefer not to implement this solution unless absolutely necessary.  If you want to implement it downstream without renaming the actual libjpeg-turbo source files, then you could simply create downstream wrappers and name them something different, e.g.:

**jcapistd-wrap.c**
```
#define BITS_IN_JSAMPLE  8
#include "jcapistd.c"
#undef BITS_IN_JSAMPLE
#define BITS_IN_JSAMPLE  12
#include "jcapistd.c"
#undef BITS_IN_JSAMPLE
#define BITS_IN_JSAMPLE  16
#include "jcapistd.c"
#undef BITS_IN_JSAMPLE
```

and have your build system include the wrappers instead of the unmodified source files.  I realize that's a hefty downstream modification, but at the moment, it doesn't make sense for me to clutter the upstream source tree unless this is an issue that really affects the ability of multiple projects to use libjpeg-turbo effectively.  I would be more receptive to such a modification in the dev/3.2 evolving branch, since all of the C source files have been moved into a **src/** subdirectory in that branch.  But again, I want to make sure that the modification is actually addressing a loss of functionality, not just addressing an innocuous warning.

**Comment by alcroito** [comment]
Hi,

Thanks for the thorough response.

I was able to reproduce the issue also with `tjbench-static`, with an unmodified stock build of the upstream sources.

```bash
$ dsymutil tjbench-static
warning: (arm64)  skipping debug map object with duplicate name and timestamp: 2024-03-15 14:07:59.000000000 /Users/alex/Dev/unsorted/libjpeg-turbo/build/libturbojpeg.a(jcapistd.c.o)
warning: (arm64)  skipping debug map object with duplicate name and timestamp: 2024-03-15 14:07:59.000000000 /Users/alex/Dev/unsorted/libjpeg-turbo/build/libturbojpeg.a(jccoefct.c.o)
warning: (arm64)  skipping debug map object with duplicate name and timestamp: 2024-03-15 14:07:59.000000000 /Users/alex/Dev/unsorted/libjpeg-turbo/build/libturbojpeg.a(jccolor.c.o)
warning: (arm64)  skipping debug map object with duplicate name and timestamp: 2024-03-15 14:07:59.000000000 /Users/alex/Dev/unsorted/libjpeg-turbo/build/libturbojpeg.a(jcdctmgr.c.o)
warning: (arm64)  skipping debug map object with duplicate name and timestamp: 2024-03-15 14:07:59.000000000 /Users/alex/Dev/unsorted/libjpeg-turbo/build/libturbojpeg.a(jcdiffct.c.o)
warning: (arm64)  skipping debug map object with duplicate name and timestamp: 2024-03-15 14:07:59.000000000 /Users/alex/Dev/unsorted/libjpeg-turbo/build/libturbojpeg.a(jclossls.c.o)
warning: (arm64)  skipping debug map object with duplicate name and timestamp: 2024-03-15 14:07:59.000000000 /Users/alex/Dev/unsorted/libjpeg-turbo/build/libturbojpeg.a(jcmainct.c.o)
warning: (arm64)  skipping debug map object with duplicate name and timestamp: 2024-03-15 14:07:59.000000000 /Users/alex/Dev/unsorted/libjpeg-turbo/build/libturbojpeg.a(jcprepct.c.o)
warning: (arm64)  skipping debug map object with duplicate name and timestamp: 2024-03-15 14:07:59.000000000 /Users/alex/Dev/unsorted/libjpeg-turbo/build/libturbojpeg.a(jcsample.c.o)
warning: (arm64)  skipping debug map object with duplicate name and timestamp: 2024-03-15 14:07:59.000000000 /Users/alex/Dev/unsorted/libjpeg-turbo/build/libturbojpeg.a(jdapistd.c.o)
warning: (arm64)  skipping debug map object with duplicate name and timestamp: 2024-03-15 14:07:59.000000000 /Users/alex/Dev/unsorted/libjpeg-turbo/build/libturbojpeg.a(jdcoefct.c.o)
warning: (arm64)  skipping debug map object with duplicate name and timestamp: 2024-03-15 14:07:59.000000000 /Users/alex/Dev/unsorted/libjpeg-turbo/build/libturbojpeg.a(jdcolor.c.o)
warning: (arm64)  skipping debug map object with duplicate name and timestamp: 2024-03-15 14:07:59.000000000 /Users/alex/Dev/unsorted/libjpeg-turbo/build/libturbojpeg.a(jddctmgr.c.o)
warning: (arm64)  skipping debug map object with duplicate name and timestamp: 2024-03-15 14:07:59.000000000 /Users/alex/Dev/unsorted/libjpeg-turbo/build/libturbojpeg.a(jddiffct.c.o)
warning: (arm64)  skipping debug map object with duplicate name and timestamp: 2024-03-15 14:07:59.000000000 /Users/alex/Dev/unsorted/libjpeg-turbo/build/libturbojpeg.a(jdlossls.c.o)
warning: (arm64)  skipping debug map object with duplicate name and timestamp: 2024-03-15 14:07:59.000000000 /Users/alex/Dev/unsorted/libjpeg-turbo/build/libturbojpeg.a(jdmainct.c.o)
warning: (arm64)  skipping debug map object with duplicate name and timestamp: 2024-03-15 14:07:59.000000000 /Users/alex/Dev/unsorted/libjpeg-turbo/build/libturbojpeg.a(jdmerge.c.o)
warning: (arm64)  skipping debug map object with duplicate name and timestamp: 2024-03-15 14:07:59.000000000 /Users/alex/Dev/unsorted/libjpeg-turbo/build/libturbojpeg.a(jdpostct.c.o)
warning: (arm64)  skipping debug map object with duplicate name and timestamp: 2024-03-15 14:07:59.000000000 /Users/alex/Dev/unsorted/libjpeg-turbo/build/libturbojpeg.a(jdsample.c.o)
warning: (arm64)  skipping debug map object with duplicate name and timestamp: 2024-03-15 14:07:59.000000000 /Users/alex/Dev/unsorted/libjpeg-turbo/build/libturbojpeg.a(jfdctfst.c.o)
warning: (arm64)  skipping debug map object with duplicate name and timestamp: 2024-03-15 14:07:59.000000000 /Users/alex/Dev/unsorted/libjpeg-turbo/build/libturbojpeg.a(jfdctint.c.o)
warning: (arm64)  skipping debug map object with duplicate name and timestamp: 2024-03-15 14:07:59.000000000 /Users/alex/Dev/unsorted/libjpeg-turbo/build/libturbojpeg.a(jidctflt.c.o)
warning: (arm64)  skipping debug map object with duplicate name and timestamp: 2024-03-15 14:07:59.000000000 /Users/alex/Dev/unsorted/libjpeg-turbo/build/libturbojpeg.a(jidctfst.c.o)
warning: (arm64)  skipping debug map object with duplicate name and timestamp: 2024-03-15 14:07:59.000000000 /Users/alex/Dev/unsorted/libjpeg-turbo/build/libturbojpeg.a(jidctint.c.o)
warning: (arm64)  skipping debug map object with duplicate name and timestamp: 2024-03-15 14:07:59.000000000 /Users/alex/Dev/unsorted/libjpeg-turbo/build/libturbojpeg.a(jidctred.c.o)
warning: (arm64)  skipping debug map object with duplicate name and timestamp: 2024-03-15 14:07:59.000000000 /Users/alex/Dev/unsorted/libjpeg-turbo/build/libturbojpeg.a(jquant1.c.o)
warning: (arm64)  skipping debug map object with duplicate name and timestamp: 2024-03-15 14:07:59.000000000 /Users/alex/Dev/unsorted/libjpeg-turbo/build/libturbojpeg.a(jquant2.c.o)
warning: (arm64)  skipping debug map object with duplicate name and timestamp: 2024-03-15 14:07:59.000000000 /Users/alex/Dev/unsorted/libjpeg-turbo/build/libturbojpeg.a(jutils.c.o)
warning: (arm64)  skipping debug map object with duplicate name and timestamp: 2024-03-15 14:07:59.000000000 /Users/alex/Dev/unsorted/libjpeg-turbo/build/libturbojpeg.a(rdppm.c.o)
warning: (arm64)  skipping debug map object with duplicate name and timestamp: 2024-03-15 14:07:59.000000000 /Users/alex/Dev/unsorted/libjpeg-turbo/build/libturbojpeg.a(wrppm.c.o)
```

In the Qt downstream build we actually get some additional warnings that I wasn't able to reproduce in the upstream build
```
warning: (arm64)  could not find object file symbol for symbol _jpeg_start_compress
warning: (arm64)  could not find object file symbol for symbol _jpeg_write_scanlines
warning: (arm64)  could not find object file symbol for symbol _jpeg_start_decompress
warning: (arm64)  could not find object file symbol for symbol _output_pass_setup
warning: (arm64)  could not find object file symbol for symbol _jpeg_read_scanlines
```

Here is the commit in `dsymutil` that added the warning. It is fairly recent:
https://github.com/llvm/llvm-project/commit/65f63739805e2c884364cb5b4b77430f96ecee7c#diff-0fd5f10e447af695bfa197b14ae4036982321a6daf41688f2b3f70587a3b9f11R262

My initial impression was: the warnings were harmless. The reason for removing the warnings was due to previous feedback from some downstream users that require explanations for each warning, think medical industry requirements, although i don't know how many of them use macOS.

The commit message i just linked implies there might be correctness issues in regards to the debug info.
The extra warnings in the Qt build might be related as well, although it's just a guess.

Reading through the dsymutil manual and the source code, it doesn't look like there is an existing option to suppress the warnings.

Having said all that, I think downstream usage in Qt will be fine with the file copying approach.

I filed the issue to see if there is any interest in using the same approach, and for search SEO purposes, in case someone else stumbles upon the issue. I realise the approach is not great, especially when developing the library itself.

If you feel like this is not very important, feel free to close the issue.

**Comment by dcommander** [comment]
OK, so you have established that it is reproducible upstream and that there are no likely workarounds. Now I just need to understand the severity of the issue, i.e. the worst-case scenario from not fixing it.

**Comment by dcommander** [comment]
If I fix it, the fix will still land in 3.1, but it should at least be back-portable into your source tree until 3.1 is available.

**Comment by alcroito** [comment]
> OK, so you have established that it is reproducible upstream and that there are no likely workarounds. Now I just need to understand the severity of the issue, i.e. the worst-case scenario from not fixing it.

Given that this is about debug symbol extraction, the worst case scenario would be missing debug info or incorrect debug info.

**Comment by alcroito** [comment]
Yes. I can reproduce it with `tjunittest-static` test, with a clean unmodified upstream `Debug` build.

If i run the test  under the debugger immediately after building it, and set a breakpoint on `jpeg_write_scanlines`, I don't get debug info.

```
$ lldb -- ./tjunittest-static
(lldb) target create "./tjunittest-static"
Current executable set to '/Users/alex/Dev/unsorted/libjpeg-turbo/build/tjunittest-static' (arm64).
(lldb) b jpeg_write_scanlines
Breakpoint 1: where = tjunittest-static`jpeg_write_scanlines, address = 0x00000001000b7228
(lldb) r
Process 55903 launched: '/Users/alex/Dev/unsorted/libjpeg-turbo/build/tjunittest-static' (arm64)
Testing 8-bit precision
Process 55903 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = breakpoint 1.1
    frame #0: 0x00000001000b7228 tjunittest-static`jpeg_write_scanlines
tjunittest-static`jpeg_write_scanlines:
->  0x1000b7228 <+0>:      sub    sp, sp, #0x30
    0x1000b722c <+103276>: stp    x29, x30, [sp, #0x20]
    0x1000b7230 <+103280>: add    x29, sp, #0x20
    0x1000b7234 <+103284>: stur   x0, [x29, #-0x8]
Target 0: (tjunittest-static) stopped.
```

If i run the same test with `./tjunittest-static -precision 12`, with an adjusted breakpoint for the different function name, i get debug info:
```
$ lldb -- ./tjunittest-static -precision 12
(lldb) target create "./tjunittest-static"
Current executable set to '/Users/alex/Dev/unsorted/libjpeg-turbo/build/tjunittest-static' (arm64).
(lldb) settings set -- target.run-args  "-precision" "12"
(lldb) b jpeg12_write_scanlines
Breakpoint 1: where = tjunittest-static`jpeg12_write_scanlines + 24 at jcapistd.c:91:7, address = 0x000000010006ead0
(lldb) r
Process 55605 launched: '/Users/alex/Dev/unsorted/libjpeg-turbo/build/tjunittest-static' (arm64)
Testing 12-bit precision
Process 55605 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = breakpoint 1.1
    frame #0: 0x000000010006ead0 tjunittest-static`jpeg12_write_scanlines(cinfo=0x000000014600c000, scanlines=0x0000000146804080, num_lines=39) at jcapistd.c:91:7
   88  	#if BITS_IN_JSAMPLE != 16 || defined(C_LOSSLESS_SUPPORTED)
   89  	  JDIMENSION row_ctr, rows_left;
   90
-> 91  	  if (cinfo->data_precision != BITS_IN_JSAMPLE)
   92  	    ERREXIT1(cinfo, JERR_BAD_PRECISION, cinfo->data_precision);
   93
   94  	  if (cinfo->global_state != CSTATE_SCANNING)
Target 0: (tjunittest-static) stopped.
```

This is all before running `dsymutil`. Running dsymutil makes no difference.

In contrast with the test that links to the shared library, i get debug info in both cases:

```
lldb -- ./tjunittest
(lldb) target create "./tjunittest"
Current executable set to '/Users/alex/Dev/unsorted/libjpeg-turbo/build/tjunittest' (arm64).
(lldb) b jpeg_write_scanlines
Breakpoint 1: where = libturbojpeg.0.dylib`jpeg_write_scanlines + 24 at jcapistd.c:91:7, address = 0x00000000000ae8ac
(lldb) r
Process 95102 launched: '/Users/alex/Dev/unsorted/libjpeg-turbo/build/tjunittest' (arm64)
Testing 8-bit precision
Process 95102 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = breakpoint 1.1
    frame #0: 0x00000001006968ac libturbojpeg.0.dylib`jpeg_write_scanlines(cinfo=0x000000015080cc00, scanlines=0x0000000150604290, num_lines=39) at jcapistd.c:91:7
   88  	#if BITS_IN_JSAMPLE != 16 || defined(C_LOSSLESS_SUPPORTED)
   89  	  JDIMENSION row_ctr, rows_left;
   90
-> 91  	  if (cinfo->data_precision != BITS_IN_JSAMPLE)
   92  	    ERREXIT1(cinfo, JERR_BAD_PRECISION, cinfo->data_precision);
   93
   94  	  if (cinfo->global_state != CSTATE_SCANNING)
Target 0: (tjunittest) stopped.
```

```
$ lldb -- ./tjunittest -precision 12
(lldb) target create "./tjunittest"
Current executable set to '/Users/alex/Dev/unsorted/libjpeg-turbo/build/tjunittest' (arm64).
(lldb) settings set -- target.run-args  "-precision" "12"
(lldb) b jpeg12_write_scanlines
Breakpoint 1: where = libturbojpeg.0.dylib`jpeg12_write_scanlines + 24 at jcapistd.c:91:7, address = 0x000000000006613c
(lldb) r
Process 54445 launched: '/Users/alex/Dev/unsorted/libjpeg-turbo/build/tjunittest' (arm64)
Testing 12-bit precision
Process 54445 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = breakpoint 1.1
    frame #0: 0x000000010064e13c libturbojpeg.0.dylib`jpeg12_write_scanlines(cinfo=0x000000014680cc00, scanlines=0x0000000146604290, num_lines=39) at jcapistd.c:91:7
   88  	#if BITS_IN_JSAMPLE != 16 || defined(C_LOSSLESS_SUPPORTED)
   89  	  JDIMENSION row_ctr, rows_left;
   90
-> 91  	  if (cinfo->data_precision != BITS_IN_JSAMPLE)
   92  	    ERREXIT1(cinfo, JERR_BAD_PRECISION, cinfo->data_precision);
   93
   94  	  if (cinfo->global_state != CSTATE_SCANNING)
Target 0: (tjunittest) stopped
```

If i do a clean build a few times, the debug info that is shown for the static lib case, shifts, sometimes it's there for `jpeg12_write_scanlines`, sometimes for `jpeg_write_scanlines`. 
Presumably because the linker chooses which object file to use non-deterministically.

**Comment by dcommander** [comment]
I can't reproduce that issue with Xcode 14.2, which implies that the issue is a regression introduced in whichever version of Xcode introduced the warning (presumably Xcode 15.)  I still posit that this is the result of the toolchain behaving incorrectly, and that opinion is bolstered by the fact that the behavior seems to have changed between Xcode 14 and 15.  I'm not sure why the tools can't look at the object file sizes if the names and timestamps are identical.  The fact that it used to work makes me wonder why it no longer does.

I looked for a possible non-disruptive workaround within CMake, such as changing the object file names for certain source files, but I couldn't find anything cleaner than your messy approach or my messy approach.  If I felt like the problem was truly in libjpeg-turbo, then I would be more inclined to work around it, but I need someone to convince me that we're doing something wrong.  I am not yet convinced of that.  Creating wrappers, as I proposed, would be messier than I thought at first.  The main problem with that approach is the need to handle duplicated file-scope definitions and static functions, and handling that would necessitate a lot of restructuring of the source files.  Before I go to that trouble, I would like to have a better understanding of why this issue only occurs with the latest Xcode releases and whether there is any potential for it to occur on other platforms.
